# Project on BRSR

This project shall take data from BRSR disclosures in India for the period of April 01, 2024 to March 31, 2025 and perform data analysis and visualization on it.

## Business Responsibility and Sustainability Reporting (BRSR)

BRSR is a sustainability reporting framework mandated by the Securities and Exchange Board of India (SEBI) for the top 1,000 listed companies in India. It requires companies to disclose their environmental, social, and governance (ESG) performance indicators, promoting responsible business practices and sustainable development. 

### Key aspects of BRSR

**Mandatory Reporting:**
BRSR is mandatory for the top 1000 listed companies in India by market capitalization.

**ESG Focus:**
It focuses on disclosing a company's performance related to Environmental, Social, and Governance factors. 

**Alignment with National Guidelines:**
BRSR disclosures are aligned with the National Guidelines on Responsible Business Conduct (NGRBC). 

**Transition from BRR:**
BRSR replaces the older Business Responsibility Reporting (BRR) framework, marking a shift towards more comprehensive and quantifiable sustainability reporting.

### Objective:
The main goal is to link a company's financial performance with its ESG performance, promoting transparency and accountability. 

# BRSR disclosures on NSE website

### Fetch NSE Data

URL: https://www.nseindia.com/companies-listing/corporate-filings-bussiness-sustainabilitiy-reports

BRSR filings for the relevant period can be downloaded as a csv file and have been saved in the project directory as **nse_brsr_filings.csv**.

In [1]:
import pandas as pd

In [2]:
#Let's take a look at what the NSE data looks like

nse_df = pd.read_csv("nse_brsr_filings.csv")
nse_df

COMPANY \n  FROM YEAR \n  TO YEAR \n  \
0                         Varun Beverages Limited          2024        2024   
1                           Castrol India Limited          2024        2024   
2                                  Cyient Limited          2023        2024   
3                                 Siemens Limited          2023        2024   
4                        Indraprastha Gas Limited          2023        2024   
...                                           ...           ...         ...   
1176                      Huhtamaki India Limited          2023        2023   
1177                         Sanofi India Limited          2023        2023   
1178  Transformers and Rectifiers (India) Limited          2023        2024   
1179                      RAIN INDUSTRIES LIMITED          2023        2023   
1180                            ABB India Limited          2023        2023   

                                          ATTACHMENT \n  \
0     https://nsearchives.nseindia.com/corporate/VBL...   
1     https://nsearchives.nseindia.com/corporate/CAS...   
2     https://nsearchives.nseindia.com/corporate/CYI...   
3     https://nsearchives.nseindia.com/corporate/SIE...   
4     https://nsearchives.nseindia.com/corporate/IGL...   
...                                                 ...   
1176  https://nsearchives.nseindia.com/corporate/HUH...   
1177  https://nsearchives.nseindia.com/corporate/SAN...   
1178  https://nsearchives.nseindia.com/corporate/TRI...   
1179  https://nsearchives.nseindia.com/corporate/RAI...   
1180  https://nsearchives.nseindia.com/corporate/ABB...   

                                              **XBRL \n  \
0     https://nsearchives.nseindia.com/corporate/xbr...   
1     https://nsearchives.nseindia.com/corporate/xbr...   
2     https://nsearchives.nseindia.com/corporate/xbr...   
3     https://nsearchives.nseindia.com/corporate/xbr...   
4     https://nsearchives.nseindia.com/corporate/xbr...   
...                                                 ...   
1176  https://nsearchives.nseindia.com/corporate/xbr...   
1177  https://nsearchives.nseindia.com/corporate/xbr...   
1178  https://nsearchives.nseindia.com/corporate/xbr...   
1179  https://nsearchives.nseindia.com/corporate/xbr...   
1180  https://nsearchives.nseindia.com/corporate/xbr...   

     ORIGINAL SUBMISSION DATE \n LATEST REVISION DATE \n  
0                    11-Mar-2025                       -  
1                    25-Feb-2025             25-Feb-2025  
2                    19-Feb-2025                       -  
3                    14-Jan-2025                       -  
4                    17-Dec-2024                       -  
...                          ...                     ...  
1176                 30-Apr-2024                       -  
1177                 23-Apr-2024                       -  
1178                 20-Apr-2024                       -  
1179                 18-Apr-2024                       -  
1180                 05-Apr-2024                       -  

[1181 rows x 7 columns]

### Clean NSE data

In [3]:
# Some basic cleaning
nse_df.rename(columns={
    "COMPANY \n": "Company",
    "FROM YEAR \n": "FromYear",
    "TO YEAR \n": "ToYear",
    "ATTACHMENT \n": "PDFURL",
    "**XBRL \n": "XBRLURL",
    "ORIGINAL SUBMISSION DATE \n": "SubmissionDate",
    "LATEST REVISION DATE \n": "RevisionDate"
}, inplace=True)

# Let's replace '-' with an empty string and convert dates to YYYY-MM-DD format
nse_df['RevisionDate'] = nse_df['RevisionDate'].replace("-", "").apply(
    lambda x: pd.to_datetime(x, format="%d-%b-%Y").strftime("%Y-%m-%d") if x else x
)

nse_df['SubmissionDate'] = nse_df['SubmissionDate'].replace("-", "").apply(
    lambda x: pd.to_datetime(x, format="%d-%b-%Y").strftime("%Y-%m-%d") if x else x
)

# Count empty values in the date columns
empty_submission = (nse_df['SubmissionDate'] == "").sum()
empty_revision = (nse_df['RevisionDate'] == "").sum()

empty_submission
empty_revision


1157

In [4]:
nse_df

Company  FromYear  ToYear  \
0                         Varun Beverages Limited      2024    2024   
1                           Castrol India Limited      2024    2024   
2                                  Cyient Limited      2023    2024   
3                                 Siemens Limited      2023    2024   
4                        Indraprastha Gas Limited      2023    2024   
...                                           ...       ...     ...   
1176                      Huhtamaki India Limited      2023    2023   
1177                         Sanofi India Limited      2023    2023   
1178  Transformers and Rectifiers (India) Limited      2023    2024   
1179                      RAIN INDUSTRIES LIMITED      2023    2023   
1180                            ABB India Limited      2023    2023   

                                                 PDFURL  \
0     https://nsearchives.nseindia.com/corporate/VBL...   
1     https://nsearchives.nseindia.com/corporate/CAS...   
2     https://nsearchives.nseindia.com/corporate/CYI...   
3     https://nsearchives.nseindia.com/corporate/SIE...   
4     https://nsearchives.nseindia.com/corporate/IGL...   
...                                                 ...   
1176  https://nsearchives.nseindia.com/corporate/HUH...   
1177  https://nsearchives.nseindia.com/corporate/SAN...   
1178  https://nsearchives.nseindia.com/corporate/TRI...   
1179  https://nsearchives.nseindia.com/corporate/RAI...   
1180  https://nsearchives.nseindia.com/corporate/ABB...   

                                                XBRLURL SubmissionDate  \
0     https://nsearchives.nseindia.com/corporate/xbr...     2025-03-11   
1     https://nsearchives.nseindia.com/corporate/xbr...     2025-02-25   
2     https://nsearchives.nseindia.com/corporate/xbr...     2025-02-19   
3     https://nsearchives.nseindia.com/corporate/xbr...     2025-01-14   
4     https://nsearchives.nseindia.com/corporate/xbr...     2024-12-17   
...                                                 ...            ...   
1176  https://nsearchives.nseindia.com/corporate/xbr...     2024-04-30   
1177  https://nsearchives.nseindia.com/corporate/xbr...     2024-04-23   
1178  https://nsearchives.nseindia.com/corporate/xbr...     2024-04-20   
1179  https://nsearchives.nseindia.com/corporate/xbr...     2024-04-18   
1180  https://nsearchives.nseindia.com/corporate/xbr...     2024-04-05   

     RevisionDate  
0                  
1      2025-02-25  
2                  
3                  
4                  
...           ...  
1176               
1177               
1178               
1179               
1180               

[1181 rows x 7 columns]

It appears that during the 01-04-2024 to 31-03-2025 period, vairous companies have filed, and for various periods. The 3 permissible periods are 
1. Calendar years 2023 (Jan to Dec 2023, submitting the report by June 2024)
2. Financial year (April 2023 to March 2024, submitting the report before Sept 2024)
3. Calendar year 2024 (Jan to Dec 2023, those submitting the report by March 2025).

Let's count the number of reports in these buckets to get a better idea.

In [5]:
count_2023_2023 = nse_df[(nse_df["FromYear"] == 2023) & (nse_df["ToYear"] == 2023)].shape[0]

count_2023_2024 = nse_df[(nse_df["FromYear"] == 2023) & (nse_df["ToYear"] == 2024)].shape[0]

count_2024_2024 = nse_df[(nse_df["FromYear"] == 2024) & (nse_df["ToYear"] == 2024)].shape[0]

# Count rows where none of conditions fits
count_neither = nse_df[~((nse_df["FromYear"] == 2023) & (nse_df["ToYear"].isin([2023, 2024]))) & 
                       ~((nse_df["FromYear"] == 2024) & (nse_df["ToYear"] == 2024))]

print(f"Rows where FromYear is 2023 and ToYear is 2023: {count_2023_2023}")
print(f"Rows where FromYear is 2023 and ToYear is 2024: {count_2023_2024}")
print(f"Rows where FromYear is 2024 and ToYear is 2024: {count_2024_2024}")
print(f"Rows where neither condition fits: {count_neither.shape[0]}")


Rows where FromYear is 2023 and ToYear is 2023: 10
Rows where FromYear is 2023 and ToYear is 2024: 1167
Rows where FromYear is 2024 and ToYear is 2024: 2
Rows where neither condition fits: 2


The cases where none of the conditions are true seem to be erroneous. Let's open them up.

In [6]:
neither_fits_df = nse_df[~((nse_df["FromYear"] == 2023) & (nse_df["ToYear"].isin([2023, 2024]))) & 
                         ~((nse_df["FromYear"] == 2024) & (nse_df["ToYear"] == 2024))]

# Display full column without truncation
pd.set_option('display.max_rows', None)  # Show all rows
pd.set_option('display.max_colwidth', None)  # Ensure full column width is displayed

neither_fits_df


Company  FromYear  ToYear  \
18                          ITI LIMITED      2024    2025   
512  Equitas Small Finance Bank Limited      2024    2025   

                                                                                                PDFURL  \
18             https://nsearchives.nseindia.com/corporate/ITI_10102024182155_SE_Intimation_on_BRSR.pdf   
512  https://nsearchives.nseindia.com/corporate/EQUITASBNK_13082024202258_ESFBBRSRIntimationFY2324.pdf   

                                                                                 XBRLURL  \
18   https://nsearchives.nseindia.com/corporate/xbrl/BRSR_1265258_10102024062627_WEB.xml   
512  https://nsearchives.nseindia.com/corporate/xbrl/BRSR_1224424_13082024083934_WEB.xml   

    SubmissionDate RevisionDate  
18      2024-10-10               
512     2024-08-13

Upon checking manually, both these cases are for FY 2023-24. So we can modify the actual df.

In [7]:
# Modify rows where FromYear = 2024 and ToYear = 2025
nse_df.loc[(nse_df["FromYear"] == 2024) & (nse_df["ToYear"] == 2025), ["FromYear", "ToYear"]] = [2023, 2024]

In [8]:
count_2023_2023 = nse_df[(nse_df["FromYear"] == 2023) & (nse_df["ToYear"] == 2023)].shape[0]

count_2023_2024 = nse_df[(nse_df["FromYear"] == 2023) & (nse_df["ToYear"] == 2024)].shape[0]

count_2024_2024 = nse_df[(nse_df["FromYear"] == 2024) & (nse_df["ToYear"] == 2024)].shape[0]

# Count rows where none of conditions fits
count_neither = nse_df[~((nse_df["FromYear"] == 2023) & (nse_df["ToYear"].isin([2023, 2024]))) & 
                       ~((nse_df["FromYear"] == 2024) & (nse_df["ToYear"] == 2024))]

print(f"Rows where FromYear is 2023 and ToYear is 2023: {count_2023_2023}")
print(f"Rows where FromYear is 2023 and ToYear is 2024: {count_2023_2024}")
print(f"Rows where FromYear is 2024 and ToYear is 2024: {count_2024_2024}")
print(f"Rows where neither condition fits: {count_neither.shape[0]}")

Rows where FromYear is 2023 and ToYear is 2023: 10
Rows where FromYear is 2023 and ToYear is 2024: 1169
Rows where FromYear is 2024 and ToYear is 2024: 2
Rows where neither condition fits: 0


### Remove duplicates

In [9]:
# Count duplicates in 'Company' column
duplicate_counts = nse_df['Company'].value_counts()
duplicates = duplicate_counts[duplicate_counts > 1]

# Display duplicate companies and their counts
print("Duplicate values in 'Company' column:")
print(duplicates)


Duplicate values in 'Company' column:
Company
Jain Irrigation Systems Limited    2
Tata Motors Limited                2
Name: count, dtype: int64


In [10]:
# Identify duplicate companies and display their rows
duplicate_companies_df = nse_df[nse_df.duplicated(subset=['Company'], keep=False)]

# Display the full DataFrame with duplicate 'Company' values
print(duplicate_companies_df)

                              Company  FromYear  ToYear  \
536   Jain Irrigation Systems Limited      2023    2024   
537   Jain Irrigation Systems Limited      2023    2024   
1139              Tata Motors Limited      2023    2024   
1140              Tata Motors Limited      2023    2024   

                                                                                            PDFURL  \
536    https://nsearchives.nseindia.com/corporate/JISLDVREQS_09082024173911_Jains_BRSR_2023_24.pdf   
537    https://nsearchives.nseindia.com/corporate/JISLJALEQS_09082024173541_Jains_BRSR_2023_24.pdf   
1139    https://nsearchives.nseindia.com/corporate/TATAMTRDVRB_31052024160028_NSEBSELETTERBRSR.pdf   
1140  https://nsearchives.nseindia.com/corporate/TATAMOTORSSJS_31052024155411_NSEBSELETTERBRSR.pdf   

                                                                                  XBRLURL  \
536   https://nsearchives.nseindia.com/corporate/xbrl/BRSR_1219094_09082024053948_WEB.xml   
537   h

These are disclosures by the same company on the same date. We will delete these 2 rows manually and keep the newer values (upper values since the original csv file was sorted in the descending order by date-time by NSE)

In [11]:
urls_to_delete = [
    "https://nsearchives.nseindia.com/corporate/xbrl/BRSR_1219073_09082024053555_WEB.xml", 
    "https://nsearchives.nseindia.com/corporate/xbrl/BRSR_1150389_31052024035442_WEB.xml"
    ]

# Remove rows where XBRLURL matches the given values
nse_df = nse_df[~nse_df["XBRLURL"].isin(urls_to_delete)]

nse_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1179 entries, 0 to 1180
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Company         1179 non-null   object
 1   FromYear        1179 non-null   int64 
 2   ToYear          1179 non-null   int64 
 3   PDFURL          1179 non-null   object
 4   XBRLURL         1179 non-null   object
 5   SubmissionDate  1179 non-null   object
 6   RevisionDate    1179 non-null   object
dtypes: int64(2), object(5)
memory usage: 73.7+ KB


In [12]:
print(nse_df.shape)

(1179, 7)


In [13]:
# Count duplicates in 'Company' column
duplicate_counts = nse_df['Company'].value_counts()
duplicates = duplicate_counts[duplicate_counts > 1]

# Display duplicate companies and their counts
print("Duplicate values in 'Company' column:")
print(duplicates)

Duplicate values in 'Company' column:
Series([], Name: count, dtype: int64)


In [14]:
nse_df.shape

(1179, 7)

### Address Submission and Revision Dates

In [15]:
# Count occurrences of each unique value in RevisionDate
revision_counts = nse_df["RevisionDate"].value_counts()
df=nse_df
sorted_df = df[df["RevisionDate"] != ""].sort_values(by="Company")
print(sorted_df)

                                    Company  FromYear  ToYear  \
666        BHARAT HEAVY ELECTRICALS LIMITED      2023    2024   
1                     Castrol India Limited      2024    2024   
1175           Craftsman Automation Limited      2023    2024   
779                             DLF Limited      2023    2024   
165    Diamond Power Infrastructure Limited      2023    2024   
690   Federal-Mogul Goetze (India) Limited.      2023    2024   
636        Flair Writing Industries Limited      2023    2024   
969             GIC Housing Finance Limited      2023    2024   
955                  GRAPHITE INDIA LIMITED      2023    2024   
841                             GRP Limited      2023    2024   
385             Gateway Distriparks Limited      2023    2024   
287          HARIOM PIPE INDUSTRIES LIMITED      2023    2024   
318                  Juniper Hotels Limited      2023    2024   
809                        KAMDHENU LIMITED      2023    2024   
88                      L

So a bunch of companies filed revised BRSR disclosures but these are not being repeated in the Company names as we saw earlier. So we downloaded them all and we consider their revision dates to be their filing dates. We will manually remove the 2 diplicate ones.

In [16]:
# Replace SubmissionDate with RevisionDate where RevisionDate is not empty
nse_df.loc[nse_df["RevisionDate"] != "", "SubmissionDate"] = nse_df["RevisionDate"]
nse_df

Company  FromYear  \
0                                       Varun Beverages Limited      2024   
1                                         Castrol India Limited      2024   
2                                                Cyient Limited      2023   
3                                               Siemens Limited      2023   
4                                      Indraprastha Gas Limited      2023   
5                                  Religare Enterprises Limited      2023   
6                                       Future Consumer Limited      2023   
7                                  JEENA SIKHO LIFECARE LIMITED      2023   
8                                          BF UTILITIES LIMITED      2023   
9                         Network18 Media & Investments Limited      2023   
10                              Procter & Gamble Health Limited      2023   
11                                       Gillette India Limited      2023   
12             Procter & Gamble Hygiene and Health Care Limited      2023   
13                        Zee Entertainment Enterprises Limited      2023   
14                                                  HMT LIMITED      2023   
15                                    THE ANDHRA SUGARS LIMITED      2023   
16                                        Mahanagar Gas Limited      2023   
17                            NETWEB TECHNOLOGIES INDIA LIMITED      2023   
18                                                  ITI LIMITED      2023   
19                                   Persistent Systems Limited      2023   
20                                          LAURUS LABS LIMITED      2023   
21                       Hindustan Construction Company Limited      2023   
22                       Mrs. Bectors Food Specialities Limited      2023   
23                                       RAJESH EXPORTS LIMITED      2023   
24                              Accelya Solutions India Limited      2023   
25                 FERTILIZERS AND CHEMICALS TRAVANCORE LIMITED      2023   
26                                      ASHIANA HOUSING LIMITED      2023   
27                      THE NEW INDIA ASSURANCE COMPANY LIMITED      2023   
28                                                 NMDC Limited      2023   
29                                           Cyient DLM Limited      2023   
30                           National Aluminium Company Limited      2023   
31                            Mazagon Dock Shipbuilders Limited      2023   
32                         Som Distilleries & Breweries Limited      2023   
33                                       PTC Industries Limited      2023   
34                                     Alicon Castalloy Limited      2023   
35                           CONFIDENCE PETROLEUM INDIA LIMITED      2023   
36                                       KITEX GARMENTS LIMITED      2023   
37                               Strides Pharma Science Limited      2023   
38                                      MUTHOOT FINANCE LIMITED      2023   
39                        TAMIL NADU NEWSPRINT & PAPERS LIMITED      2023   
40                              Savita Oil Technologies Limited      2023   
41                                      RHI Magnesita India Ltd      2023   
42                                           NMDC STEEL LIMITED      2023   
43                        Entertainment Network (India) Limited      2023   
44                             CHAMAN LAL SETIA EXPORTS LIMITED      2023   
45                  Nagarjuna Fertilizers and Chemicals Limited      2023   
46                                   GENSOL ENGINEERING LIMITED      2023   
47                                    Jindal Poly Films Limited      2023   
48                                          PC Jeweller Limited      2023   
49                                    APEX FROZEN FOODS LIMITED      2023   
50                           Salasar Techno Engineering Limited      2023   
51                                  NAZARA TECHNOLOG

In [17]:
# Finally, let's drop the RevisionDate column
nse_df.drop(columns=["RevisionDate"], inplace=True)

nse_df

C:\Users\manoj\AppData\Local\Temp\ipykernel_18176\761217088.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  nse_df.drop(columns=["RevisionDate"], inplace=True)


Company  FromYear  \
0                                       Varun Beverages Limited      2024   
1                                         Castrol India Limited      2024   
2                                                Cyient Limited      2023   
3                                               Siemens Limited      2023   
4                                      Indraprastha Gas Limited      2023   
5                                  Religare Enterprises Limited      2023   
6                                       Future Consumer Limited      2023   
7                                  JEENA SIKHO LIFECARE LIMITED      2023   
8                                          BF UTILITIES LIMITED      2023   
9                         Network18 Media & Investments Limited      2023   
10                              Procter & Gamble Health Limited      2023   
11                                       Gillette India Limited      2023   
12             Procter & Gamble Hygiene and Health Care Limited      2023   
13                        Zee Entertainment Enterprises Limited      2023   
14                                                  HMT LIMITED      2023   
15                                    THE ANDHRA SUGARS LIMITED      2023   
16                                        Mahanagar Gas Limited      2023   
17                            NETWEB TECHNOLOGIES INDIA LIMITED      2023   
18                                                  ITI LIMITED      2023   
19                                   Persistent Systems Limited      2023   
20                                          LAURUS LABS LIMITED      2023   
21                       Hindustan Construction Company Limited      2023   
22                       Mrs. Bectors Food Specialities Limited      2023   
23                                       RAJESH EXPORTS LIMITED      2023   
24                              Accelya Solutions India Limited      2023   
25                 FERTILIZERS AND CHEMICALS TRAVANCORE LIMITED      2023   
26                                      ASHIANA HOUSING LIMITED      2023   
27                      THE NEW INDIA ASSURANCE COMPANY LIMITED      2023   
28                                                 NMDC Limited      2023   
29                                           Cyient DLM Limited      2023   
30                           National Aluminium Company Limited      2023   
31                            Mazagon Dock Shipbuilders Limited      2023   
32                         Som Distilleries & Breweries Limited      2023   
33                                       PTC Industries Limited      2023   
34                                     Alicon Castalloy Limited      2023   
35                           CONFIDENCE PETROLEUM INDIA LIMITED      2023   
36                                       KITEX GARMENTS LIMITED      2023   
37                               Strides Pharma Science Limited      2023   
38                                      MUTHOOT FINANCE LIMITED      2023   
39                        TAMIL NADU NEWSPRINT & PAPERS LIMITED      2023   
40                              Savita Oil Technologies Limited      2023   
41                                      RHI Magnesita India Ltd      2023   
42                                           NMDC STEEL LIMITED      2023   
43                        Entertainment Network (India) Limited      2023   
44                             CHAMAN LAL SETIA EXPORTS LIMITED      2023   
45                  Nagarjuna Fertilizers and Chemicals Limited      2023   
46                                   GENSOL ENGINEERING LIMITED      2023   
47                                    Jindal Poly Films Limited      2023   
48                                          PC Jeweller Limited      2023   
49                                    APEX FROZEN FOODS LIMITED      2023   
50                           Salasar Techno Engineering Limited      2023   
51                                  NAZARA TECHNOLOG

Let's save this DataFrame into an Excel file so that we can pick up from this point in the next notebook.

### Download NSE data

We will take the XBRL URL, save the XML file, convert it to Excel using the template available on NSE website and save those excel files as well.

In [18]:
import pandas as pd
from tqdm import tqdm
from pathlib import Path
import requests
import xml.etree.ElementTree as ET
from openpyxl import Workbook, load_workbook
from openpyxl.styles import Font
from openpyxl.utils import get_column_letter

# === CONFIGURATION ===
TEMPLATE_PATH = "template.xlsx"
OUTPUT_DIR = Path("excel_files")
XML_DIR = Path("xbrl_files")
OUTPUT_DIR.mkdir(exist_ok=True)
XML_DIR.mkdir(exist_ok=True)

# === CONSTANTS ===
NS_XBRLI = "{http://www.xbrl.org/2003/instance}"

# === HELPER FUNCTIONS ===
def concepts_from_template(template_path: str) -> set[str]:
    wb = load_workbook(template_path, read_only=True)
    ws = wb["Element"]
    header = [c.value for c in next(ws.iter_rows(max_row=1))]
    idx = header.index("name")
    concepts = {str(row[idx]).strip() for row in ws.iter_rows(min_row=2, values_only=True) if row[idx]}
    wb.close()
    return concepts

def load_context_periods(root) -> dict[str, str]:
    ctx_period = {}
    for ctx in root.findall(NS_XBRLI + "context"):
        ctx_id = ctx.attrib.get("id")
        period = ctx.find(NS_XBRLI + "period")
        if not period:
            continue
        sd = period.find(NS_XBRLI + "startDate")
        ed = period.find(NS_XBRLI + "endDate")
        inst = period.find(NS_XBRLI + "instant")
        if sd is not None and ed is not None:
            ctx_period[ctx_id] = f"{sd.text} To {ed.text}"
        elif inst is not None:
            ctx_period[ctx_id] = inst.text
        else:
            ctx_period[ctx_id] = ""
    return ctx_period

def extract_facts(xbrl_bytes: bytes, wanted_concepts: set[str]):
    root = ET.fromstring(xbrl_bytes)
    ctx_period = load_context_periods(root)
    for elem in root:
        tag = elem.tag
        if tag.endswith(("context", "unit", "schemaRef")):
            continue
        local = tag.split("}", 1)[1] if "}" in tag else tag
        if local not in wanted_concepts:
            continue
        ctx_id = elem.attrib.get("contextRef")
        period = ctx_period.get(ctx_id, "") if ctx_id else ""
        decimals = elem.attrib.get("decimals")
        value = (elem.text or "").strip()
        yield local, period, ctx_id or None, decimals, value

def build_workbook(facts) -> Workbook:
    wb = Workbook()
    ws = wb.active
    ws.title = "Instance Data"
    ws.append(["Sr.No.", "Element Name", "Period", "Unit", "Decimals", "Fact Value"])
    for cell in ws[1]:
        cell.font = Font(bold=True)
    for sr, (local, period, ctx, dec, val) in enumerate(facts, start=1):
        ws.append([sr, local, period, ctx, dec, val])
    for col in range(1, 7):
        ws.column_dimensions[get_column_letter(col)].width = 25
    return wb

# === MAIN FUNCTION (NOW WITH XML DOWNLOAD) ===
def xml_url_to_excel(xml_url: str, company_name: str, template_path: str = TEMPLATE_PATH) -> str:
    try:
        headers = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
        }

        # Save path for XML
        xml_path = XML_DIR / f"{company_name}.xml"

        # Skip download if file already exists
        if not xml_path.exists():
            response = requests.get(xml_url, headers=headers, timeout=30)
            response.raise_for_status()
            with open(xml_path, "wb") as f:
                f.write(response.content)
        else:
            response = open(xml_path, "rb")

        # Load template concepts
        concepts = concepts_from_template(template_path)

        # Read the XBRL bytes
        xbrl_bytes = open(xml_path, "rb").read()

        # Extract facts
        facts = list(extract_facts(xbrl_bytes, concepts))
        if not facts:
            return None

        # Create Excel file
        wb = build_workbook(facts)
        excel_path = OUTPUT_DIR / f"{company_name.replace(' ', '_')}_InstanceData.xlsx"
        wb.save(excel_path)
        return str(excel_path)

    except Exception as e:
        print(f"❌ Error processing {company_name} | {xml_url}: {e}")
        return None


In [19]:
missing_files = []

for _, row in tqdm(nse_df.iterrows(), total=nse_df.shape[0], desc="Processing XBRL Files"):
    xml_url = row["XBRLURL"]
    company_name = row["Company"].strip()

    if pd.notna(xml_url) and xml_url.strip():
        result = xml_url_to_excel(xml_url, company_name)
        if result is None:
            missing_files.append(f"{company_name} | {xml_url}")

if missing_files:
    print("\n❌ The following files failed:")
    for entry in missing_files:
        print(entry)

print("✅ All done!")


Processing XBRL Files:   1%|▎                                                         | 6/1179 [00:11<38:09,  1.95s/it]


KeyboardInterrupt: 

So there are the 6 cases where the XML file URLs are absent or broken. We will remove those from our DF.

In [20]:
# List of companies you want to filter
companies_to_filter = ["GMR Power and Urban Infra Limited", "GE T&D India Limited", "Amara Raja Energy & Mobility Limited", 
                       "Mahindra & Mahindra Limited", "Mahindra & Mahindra Financial Services Limited"]

# Filter rows where Company is in the list
filtered_df = nse_df[nse_df["Company"].isin(companies_to_filter)]

# Display filtered DataFrame
print("Matching rows:")
print(filtered_df)

Matching rows:
                                             Company  FromYear  ToYear  \
407                GMR Power and Urban Infra Limited      2023    2024   
518                             GE T&D India Limited      2023    2024   
872             Amara Raja Energy & Mobility Limited      2023    2024   
998                      Mahindra & Mahindra Limited      2023    2024   
1045  Mahindra & Mahindra Financial Services Limited      2023    2024   

                                                                                                   PDFURL  \
407                        https://nsearchives.nseindia.com/corporate/GPUIL_24082024232635_BRSRattach.pdf   
518                            https://nsearchives.nseindia.com/corporate/GETDIL1_13082024165119_BRSR.pdf   
872                    https://nsearchives.nseindia.com/corporate/AMARAJABAT_10072024190429_BRSR_2024.pdf   
998        https://nsearchives.nseindia.com/corporate/ferozebaria_29062024195438_AnnualReportwithBRSR.pd

In [21]:
# Remove rows where Company matches any value in the list
final_df = nse_df[~nse_df["Company"].isin(companies_to_filter)]

# Display updated DataFrame
print("✅ Rows deleted. Updated DataFrame:")
print(final_df)


✅ Rows deleted. Updated DataFrame:
                                                        Company  FromYear  \
0                                       Varun Beverages Limited      2024   
1                                         Castrol India Limited      2024   
2                                                Cyient Limited      2023   
3                                               Siemens Limited      2023   
4                                      Indraprastha Gas Limited      2023   
5                                  Religare Enterprises Limited      2023   
6                                       Future Consumer Limited      2023   
7                                  JEENA SIKHO LIFECARE LIMITED      2023   
8                                          BF UTILITIES LIMITED      2023   
9                         Network18 Media & Investments Limited      2023   
10                              Procter & Gamble Health Limited      2023   
11                                       

In [22]:
final_df.shape

(1174, 6)

In [23]:
final_df.head()

Company  FromYear  ToYear  \
0   Varun Beverages Limited      2024    2024   
1     Castrol India Limited      2024    2024   
2            Cyient Limited      2023    2024   
3           Siemens Limited      2023    2024   
4  Indraprastha Gas Limited      2023    2024   

                                                                                                         PDFURL  \
0                             https://nsearchives.nseindia.com/corporate/VBL_11032025174358_VBLNoticeAR2024.pdf   
1  https://nsearchives.nseindia.com/corporate/CASTROLINDIA_25022025195909_SE_Intimation_BRSRFY2024_25022025.pdf   
2                                 https://nsearchives.nseindia.com/corporate/CYIENT_19022025125655_BRSR2024.pdf   
3             https://nsearchives.nseindia.com/corporate/SIEMENS57_14012025173818_Siemens-Limited-BRSR-2024.pdf   
4                                    https://nsearchives.nseindia.com/corporate/IGL1_17122024163450_BRSRIGL.pdf   

                                                                               XBRLURL  \
0  https://nsearchives.nseindia.com/corporate/xbrl/BRSR_1394994_11032025054427_WEB.xml   
1  https://nsearchives.nseindia.com/corporate/xbrl/BRSR_1389867_25022025075940_WEB.xml   
2  https://nsearchives.nseindia.com/corporate/xbrl/BRSR_1387421_19022025125706_WEB.xml   
3  https://nsearchives.nseindia.com/corporate/xbrl/BRSR_1345203_14012025053859_WEB.xml   
4  https://nsearchives.nseindia.com/corporate/xbrl/BRSR_1328008_17122024043540_WEB.xml   

  SubmissionDate  
0     2025-03-11  
1     2025-02-25  
2     2025-02-19  
3     2025-01-14  
4     2024-12-17

So we're left with 1181 - 2 (duplicate) - 5 (broken links) = 1174 rows

### Save final DF to Excel

In [24]:
final_df.to_excel("nse_data.xlsx", index=False)

In [ ]:
import os

# Define the path to the folder containing your Excel files
folder_path = r"excel_files"

# Loop over all files in the folder
for filename in os.listdir(folder_path):
    if filename.endswith("_InstanceData.xlsx"):
        # Create the new filename by removing '_InstanceData'
        new_filename = filename.replace("_InstanceData.xlsx", ".xlsx")
        # Build full file paths
        src = os.path.join(folder_path, filename)
        dst = os.path.join(folder_path, new_filename)
        # Rename the file
        os.rename(src, dst)
        print(f"Renamed: {filename} ➜ {new_filename}")
